package NKikimrFileStore;
option java_package = "ru.yandex.kikimr.proto";

message TChannelProfile {
    // storage settings
    optional string PoolKind = 1;
    optional uint64 Size = 2; // size in bytes
    optional double ReadIops = 3; // required iops for read requests
    optional double ReadBandwidth = 4; // required bandwidth (in bytes/sec) for read requests
    optional double WriteIops = 5; // required iops for write requests
    optional double WriteBandwidth = 6; // required bandwidth (in bytes/sec) for write requests
    optional uint32 DataKind = 7; // the kind of data to be stored, e.g. mixed/merged/etc.
}

message TConfig {
    // Version number for atomic updates
    optional uint32 Version = 1;

    // FileStore information
    optional string FileSystemId = 10;
    optional string ProjectId = 11;
    optional string FolderId = 12;
    optional string CloudId = 13;

    optional uint32 BlockSize = 14;
    optional uint64 BlocksCount = 15;

    optional uint64 CreationTs = 16;
    optional uint64 AlterTs = 17;

    optional uint32 RangeIdHasherType = 18;

    repeated TChannelProfile ExplicitChannelProfiles = 30;
    optional bool PoolKindChangeAllowed = 31;

    optional uint32 StorageMediaKind = 32;

    optional uint64 PerformanceProfileMaxReadBandwidth = 33;
    optional uint32 PerformanceProfileMaxReadIops = 34;
    optional uint64 PerformanceProfileMaxWriteBandwidth = 35;
    optional uint32 PerformanceProfileMaxWriteIops = 36;
    optional uint32 PerformanceProfileBoostTime = 37;
    optional uint32 PerformanceProfileBoostRefillTime = 38;
    optional uint32 PerformanceProfileBoostPercentage = 39;

    optional uint32 NodesCount = 40;

    optional bool PerformanceProfileThrottlingEnabled = 41;
    optional uint32 PerformanceProfileBurstPercentage = 42;
    optional uint32 PerformanceProfileDefaultPostponedRequestWeight = 43;
    optional uint32 PerformanceProfileMaxPostponedWeight = 44;
    optional uint32 PerformanceProfileMaxWriteCostMultiplier = 45;
    optional uint32 PerformanceProfileMaxPostponedTime = 46;
    optional uint32 PerformanceProfileMaxPostponedCount = 47;

    // Indicates that filestore does not belong to user directly, but used for system needs
    optional bool IsSystem = 48;
}

message TUpdateConfig {
    optional uint64 TxId = 1;
    optional TConfig Config = 2;
}

enum EStatus {
    OK = 0;
    ERROR = 1;
    ERROR_BAD_VERSION = 2;
    ERROR_UPDATE_IN_PROGRESS = 3;
}

message TUpdateConfigResponse {
    optional uint64 TxId = 1;
    optional uint64 Origin = 2;
    optional EStatus Status = 3;
}
